airbyte-cdk/python/airbyte_cdk/sources/connector_state_manager.py

```diff
@@ -73,6 +74,31 @@ def update_state_for_stream(self, stream_name: str, namespace: Optional[str], va
         stream_descriptor = HashableStreamDescriptor(name=stream_name, namespace=namespace)
         self.per_stream_states[stream_descriptor] = AirbyteStateBlob.parse_obj(value)
 
+    def create_state_message(self, stream_name: str, namespace: Optional[str]) -> AirbyteMessage:
+        """
+        Generates an AirbyteMessage using the current per-stream state of a specified stream
+        :param stream_name: The name of the stream for the message that is being created
+        :param namespace: The namespace of the stream for the message that is being created
+        :return: The Airbyte state message to be emitted by the connector during a sync
+        """
+        hashable_descriptor = HashableStreamDescriptor(name=stream_name, namespace=namespace)
+        stream_state = self.per_stream_states.get(hashable_descriptor, AirbyteStateBlob())
+        if namespace is None:
+            # In the Airbyte protocol, the StreamDescriptor namespace field is not required. However, the platform will throw a
+            # validation error if it receives namespace=null. That is why if namespace is None, the field should be omitted instead.
+            stream_descriptor = StreamDescriptor(name=stream_name)
```

Comment:
weird, i don't know why this doesn't show up inline, but its correctly shown in the main PR page. reposting

I was doing some final testing on my local airbyte and I found a surprisingly dangerous gotcha. The platform allows for either a string or no field to be specified at all, but we CANNOT send `namespace=None` to the platform. If we do, the platform will fail validation and state will not get saved to the DB.
```
airbyte-worker      | 2022-09-15 04:48:20 INFO i.a.v.j.JsonSchemaValidator(test):71 - JSON schema validation failed. 
airbyte-worker      | errors: $.state.stream.stream_descriptor.namespace: null found, string expected
airbyte-worker      | 2022-09-15 04:48:20 ERROR i.a.w.i.DefaultAirbyteStreamFactory(lambda$create$1):71 - Validation failed: {"type":"STATE","state":{"type":"STREAM","stream":{"stream_descriptor":{"name":"blocks","namespace":null},"stream_state":{}},"data":{"blocks":{}}}}
```

Due to how pydantic is designed, it is really hard to validate in a test that the field is “unset”, so we can’t really test for this at least from what i tried. For the pydantic model, regardless of whether it is actually unset or if it was set to none, asserting against the object always yields None. And one careless mistake of adding the field back in here will completely break incremental checkpointing.

I think this is worth at least a discussion, but I'm on the side of always emitting and therefore receiving back the empty string for the `namespace` field so we don't run the risk of accidentally passing `namespace=null`. And this can at the very least allow for tests to be written to cover that scenario.

Reply:
nvm found a way to check for unset fields with pydantic, we should be okay on this since I can test for this now